src/Renci.SshNet/PrivateKeyFile.cs

```diff
@@ -518,38 +561,76 @@ private static Key ParseOpenSshV1Key(byte[] keyFileData, string passPhrase)
+                    case "aes128-cbc":
+                        cipherInfo = new CipherInfo(128, (key, iv) => new AesCipher(key, iv, AesCipherMode.CBC, pkcs7Padding: false));
+                        break;
+                    case "aes192-cbc":
+                        cipherInfo = new CipherInfo(192, (key, iv) => new AesCipher(key, iv, AesCipherMode.CBC, pkcs7Padding: false));
+                        break;
                     case "aes256-cbc":
-                        cipher = new AesCipher(key, iv, AesCipherMode.CBC, pkcs7Padding: false);
+                        cipherInfo = new CipherInfo(256, (key, iv) => new AesCipher(key, iv, AesCipherMode.CBC, pkcs7Padding: false));
+                        break;
+                    case "aes128-ctr":
+                        cipherInfo = new CipherInfo(128, (key, iv) => new AesCipher(key, iv, AesCipherMode.CTR, pkcs7Padding: false));
+                        break;
+                    case "aes192-ctr":
+                        cipherInfo = new CipherInfo(192, (key, iv) => new AesCipher(key, iv, AesCipherMode.CTR, pkcs7Padding: false));
                         break;
                     case "aes256-ctr":
-                        cipher = new AesCipher(key, iv, AesCipherMode.CTR, pkcs7Padding: false);
+                        cipherInfo = new CipherInfo(256, (key, iv) => new AesCipher(key, iv, AesCipherMode.CTR, pkcs7Padding: false));
+                        break;
+                    case "aes128-gcm@openssh.com":
+                        cipherInfo = new CipherInfo(128, (key, iv) => new AesGcmCipher(key, iv, aadLength: 0), isAead: true);
+                        break;
+                    case "aes256-gcm@openssh.com":
+                        cipherInfo = new CipherInfo(256, (key, iv) => new AesGcmCipher(key, iv, aadLength: 0), isAead: true);
+                        break;
+                    case "chacha20-poly1305@openssh.com":
+                        ivLength = 12;
+                        cipherInfo = new CipherInfo(256, (key, iv) => new ChaCha20Poly1305Cipher(key, aadLength: 0), isAead: true);
                         break;
                     default:
                         throw new SshException("Cipher '" + cipherName + "' is not supported for an OpenSSH key.");
                 }
 
+                var keyLength = cipherInfo.KeySize / 8;
+
+                // inspired by the SSHj library (https://github.com/hierynomus/sshj)
+                // apply the kdf to derive a key and iv from the passphrase
+                var passPhraseBytes = Encoding.UTF8.GetBytes(passPhrase);
+                var keyiv = new byte[keyLength + ivLength];
+                new BCrypt().Pbkdf(passPhraseBytes, salt, rounds, keyiv);
+
+                var key = keyiv.Take(keyLength);
+                var iv = keyiv.Take(keyLength, ivLength);
+
+                var cipher = cipherInfo.Cipher(key, iv);
+                var cipherData = keyReader.ReadBytes(privateKeyLength + cipher.TagSize);
```

Comment:
Can you point me to where the `+ TagSize` comes from?

Reply:
Not quite sure if I understand your question... The AEAD ciphers will generate tag when encrypt data. In this case (OpenSSH), the tag is appended to the cipher text. When decrypt and verify authenticity, we need to read not only cipher data but also the tag.

Reply:
I am asking where you have found that the tag is appended to the cipher data. In https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.key I just see `string	encrypted, padded list of private keys`

Reply:
https://github.com/openssh/openssh-portable/blob/master/sshkey.c#L2951

Reply:
Thanks

Reply:
And https://github.com/openssh/openssh-portable/blob/master/cipher.c#L340